apisix/plugins/hmac-auth.lua

```diff
@@ -194,14 +209,28 @@ local function generate_signature(ctx, secret_key, params)
         end
         core.table.sort(keys)
 
+        local field_val = get_conf_field(params.access_key, "enable_encode")
+        core.log.info("enable_encode: ", field_val)
+
         for _, key in pairs(keys) do
             local param = args[key]
-            if type(param) == "table" then
-                for _, val in pairs(param) do
-                    core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(val))
+            -- whether to escape the uri parameters
+            if field_val then
```

Comment:
Would be better if we use an empty function to escape the uri params when the option is disable. Like, `core.table.insert(query_tab, encode_or_not(key) .. "=" .. encode_or_not(val))`. When you do it, note that, please avoid creating a new function per request.

Reply:
@spacewander 
You mean we need to use a function like this: `local function encode_or_not(key) return key end` . To handle the situation where this option is disabled?

Reply:
`encode_or_not` is a variable, like that: `if xxx then encode_or_not = encode else encode_or_not = do_nothing`.

Reply:
@spacewander  PTAL. I am not sure if the modification here is correct.

Reply:
Actually I mean this change:
```diff
diff --git apisix/plugins/hmac-auth.lua apisix/plugins/hmac-auth.lua
index bc09d21..45cc0cc 100644
--- apisix/plugins/hmac-auth.lua
+++ apisix/plugins/hmac-auth.lua
@@ -190,6 +190,11 @@ local function get_conf_field(access_key, field_name)
 end


+local function do_nothing(v)
+    return v
+end
+
+
 local function generate_signature(ctx, secret_key, params)
     local canonical_uri = ctx.var.uri
     local canonical_query_string = ""
@@ -212,25 +217,22 @@ local function generate_signature(ctx, secret_key, params)
         local encode_or_not = get_conf_field(params.access_key, "encode_uri_params")
         core.log.info("encode_uri_params: ", encode_or_not)

+        local encode_uri_param = do_nothing
+        if encode_or_not then
+            encode_uri_param = escape_uri
+        end
+
         for _, key in pairs(keys) do
             local param = args[key]
             -- whether to encode the uri parameters
-            if encode_or_not then
-                if type(param) == "table" then
-                    for _, val in pairs(param) do
-                        core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(val))
-                    end
-                else
-                    core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(param))
+            if type(param) == "table" then
+                for _, val in pairs(param) do
+                    core.table.insert(query_tab, encode_uri_param(key) .. "="
+                                      .. encode_uri_param(val))
                 end
             else
-                if type(param) == "table" then
-                    for _, val in pairs(param) do
-                        core.table.insert(query_tab, key .. "=" .. val)
-                    end
-                else
-                    core.table.insert(query_tab, key .. "=" .. param)
-                end
+                core.table.insert(query_tab, encode_uri_param(key) .. "="
+                                  .. encode_uri_param(param))
             end
         end
         canonical_query_string = core.table.concat(query_tab, "&")
```

You can integrate it into your code.

Reply:
ok, thank you very much.